modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Enrich.scala

```diff
@@ -0,0 +1,147 @@
+
+import fs2.Stream
+
+import _root_.io.sentry.SentryClient
+
+import _root_.io.circe.Json
+import _root_.io.circe.syntax._
+
+import _root_.io.chrisdavenport.log4cats.Logger
+import _root_.io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+
+import com.snowplowanalytics.iglu.client.Client
+
+import com.snowplowanalytics.snowplow.badrows.{Processor, BadRow, Failure, Payload => BadRowPayload}
+
+import com.snowplowanalytics.snowplow.enrich.common.EtlPipeline
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+import com.snowplowanalytics.snowplow.enrich.common.adapters.AdapterRegistry
+import com.snowplowanalytics.snowplow.enrich.common.loaders.{CollectorPayload, ThriftLoader}
+import com.snowplowanalytics.snowplow.enrich.common.enrichments.EnrichmentRegistry
+
+object Enrich {
+
+  /** Default adapter registry, can be constructed dynamically in future */
+  val adapterRegistry = new AdapterRegistry()
+
+  val processor: Processor = Processor(generated.BuildInfo.name, generated.BuildInfo.version)
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  /**
+   * Run a primary enrichment stream, reading from [[Environment]] source, enriching
+   * via [[enrichWith]] and sinking into [[GoodSink]] and [[BadSink]] respectively.
+   * Can be stopped via _stop signal_ from [[Environment]]
+   *
+   * The stream won't download any enrichment DBs. [[Assets]] stream or
+   * [[Assets.download]] should be responsible for that
+   */
+  def run[F[_]: Concurrent: ContextShift: Clock](env: Environment[F]): Stream[F, Unit] = {
+    val registry: F[EnrichmentRegistry[F]] = env.enrichments.get.map(_.registry)
+    val enrich: Enrich[F] = enrichWith[F](registry, env.igluClient, env.sentry, env.metrics.enrichLatency)
+    val badSink: BadSink[F] = _.evalTap(_ => env.metrics.badCount).through(env.bad)
+    val goodSink: GoodSink[F] = _.evalTap(_ => env.metrics.goodCount).through(env.good)
+
+    env.source
+      .pauseWhen(env.pauseEnrich)
+      .evalTap(_ => env.metrics.rawCount)
+      .parEvalMapUnordered(16)(payload => env.blocker.blockOn(enrich(payload)))
+      .flatMap(_.decompose[BadRow, EnrichedEvent])
+      .observeEither(badSink, goodSink)
+      .void
+  }
+
+  /**
+   * Enrich a single [[CollectorPayload]] to get list of bad rows and/or enriched events
+   *
+   * Along with actual `ack` the `enrichLatency` gauge will be updated
+   */
+  def enrichWith[F[_]: Clock: Sync](
+    enrichRegistry: F[EnrichmentRegistry[F]],
+    igluClient: Client[F, Json],
+    sentry: Option[SentryClient],
+    enrichLatency: Option[Long] => F[Unit]
+  )(
+    row: Payload[F, Array[Byte]]
+  ): F[Result[F]] = {
+    val payload = ThriftLoader.toCollectorPayload(row.data, processor)
+    val collectorTstamp = payload.toOption.flatMap(_.flatMap(_.context.timestamp).map(_.getMillis))
+
+    val result =
+      for {
+        _ <- Logger[F].debug(payloadToString(payload))
```

Comment:
Yes, I do. It's obviously for super rare situations where we can print the data being processed, but in case we'll need this again - we won't have to release a new asset for debugging only purposes.